java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -159,198 +106,162 @@ class EntityHandleManager(private val handleManager: HandleManager) {
     }
 }
 
-internal open class HandleEventBase<T, H : Handle> {
-    protected val onUpdateActions: MutableList<(T) -> Unit> = mutableListOf()
-    protected val onSyncActions: MutableList<(H) -> Unit> = mutableListOf()
-    protected val onDesyncActions: MutableList<(H) -> Unit> = mutableListOf()
+/** Base functionality common to all read/write singleton and collection handles. */
+sealed class HandleAdapter(
+    override val mode: HandleMode,
+    override val name: String
+) : Handle {
+    private val onSyncActions: MutableList<() -> Unit> = mutableListOf()
+    private val onDesyncActions: MutableList<() -> Unit> = mutableListOf()
 
-    suspend fun onUpdate(action: (T) -> Unit) {
-        onUpdateActions.add(action)
+    override fun onSync(action: () -> Unit) {
+        onSyncActions.add(action)
     }
 
-    suspend fun onSync(action: (H) -> Unit) {
+    override fun onDesync(action: () -> Unit) {
         onSyncActions.add(action)
     }
 
-    suspend fun onDesync(action: (H) -> Unit) {
-        onDesyncActions.add(action)
+    protected fun fireSync() {
+        onSyncActions.forEach { it() }
     }
 
-    protected suspend fun fireUpdate(value: T) {
-        onUpdateActions.forEach { action -> action(value) }
+    protected fun fireDesync() {
+        onDesyncActions.forEach { it() }
     }
 
-    protected suspend fun fireSync() {
-        onSyncActions.forEach { action -> action(this as H) }
+    protected fun checkCanRead() {
+        if (!mode.canRead) {
+            throw IllegalArgumentException("Handle does not support reads.")
```

Comment:
Handle $name does not support reads. That'll be super-useful in logs.

Reply:
Done